README
#==============================================================
#
#  Name: README
#
#  Rem:  	This is a driver for the Xycom 2445 32-channel optically
#		isolated digital output Industry Pack (IP) with interrupts.
#		This driver also supports the Acromag IP445
#		series digital input cards.
#
#		In order to deploy in a uniform fashion w/ the many other ip* modules,
#		this module has been renamed from xy2445 to ip445.
#		Support for the Xycom db, driver, and function names are all preserved
#		for compatibility, but ip445 variants have also been made available.
#
#
#  Auth: dd-mmm-yyyy, Andy Foster        (USERNAME)
#  Rev:  dd-mmm-yyyy, Reviewer's Name    (USERNAME)
#--------------------------------------------------------------
#  Mod:
#        05-May-2009, S. Peng            (pengs):
#          Comments out readback during init to help restore
#        29-Apr-2009, K. Luchini         (LUCHINI):
#          Now built under RTEMS-4.9.1 and EPICS-R3.14.8.2
#
#==============================================================
#
EPICS Verions:
--------------
This driver has currently been built and tested at SLAC with EPICS 3.14.8.2.
EPICS OSI support has been added to this driver by Till Strauman at SLAC.

Operating System:
-------------------
This driver has only been tested under RTEMS since osi support has been added.

How to use this module in your IOC
------------------
1.      Add the following lines in your iocTop/configure/RELEASE, using your desired IP445_MODULE_VERSION:
IP445_MODULE_VERSION=R2.3.0
IP445=$(EPICS_MODULES)/ip445/$(IP445_MODULE_VERSION)

2.
For XyCom 2445, copy the following file to your iocTop/<iocName>/<iocName>App/Db directory:
        xy2445App/Db/xy2445.substitutions
For Acromag IP445, copy the following file to your iocTop/<iocName>/<iocName>App/Db directory:
        xy2445App/Db/ip445.substitutions
Customize as needed for your IOC

3.
For XyCom 2445, add the following lines in your iocTop/<iocName>/<iocName>App/Db/Makefile:
DB += xy2445.db
For Acromag IP445, add the following lines in your iocTop/<iocName>/<iocName>App/Db/Makefile:
DB += ip445.db

4.      Add the following lines in your iocTop/<iocName>/<iocName>App/src/Makefile:
Note: Use either the ip445 variant or the Xy2445 variant here, but NOT both!
<iocName>_DBD += ip445.dbd
<iocName>_LIBS += ip445
   or
<iocName>_DBD += Xy2445.dbd
<iocName>_LIBS += Xy2445

Add the following lines in your iocTop/<iocName>/iocBoot/ioc<iocName>/st.cmd,
selecting the Xycom or Acromag as needed.
# Initialize IPAC carrier
# # The 2nd parameter is the base I/O address set by the jumpers on your carrier board
# # Make sure this address doesn't conflict with other boards.
# # Note: ipac supports other carrier cards besides xy9660.
# # See the ipac_carrier_t types in the ipac module for other options.
# ipacAddCarrier(&xy9660,"0x0000")
#
# # Initialize digital input module
# # Set card to the carrier number, which is 0 for the first ipacAddCarrier() call and
# # increments as other carrier cards are added.
# # The slot numbers are 0..3 corresponding to slots A..D on the carrier card
# # More than one digital input card can be used as long as the name and slot # differ.
# # ip445Create( <name>, <card>, <slot>, <modeName>, <intHndlr>, <usrFunc>, <vector>, <event>, <debounce> )
# ip445Create( "di0", 0, 0, "STANDARD", "LEVEL", 0x0, 0x80, 0x0, 0x00 )
# # or for backward compatibility, you can use xy2445Create.  One call per card
# xy2445Create( "di0", 0, 0, "STANDARD", "LEVEL", 0x0, 0x80, 0x0, 0x00 )
#
# # Load the EPICS records
# # For XyCom 2445:
# dbLoadRecords("db/xy2445.db","CARD=di0")
# # For Acromag IP445:
# # dbLoadRecords("db/ip445.db","CARD=di0")
#
# ####################################################

EPICS PV Values:
----------------------
The default names assigned by the db template and substitution files
in this module are shown below.  These can be customized after you copy the
template and substitution files to your ioc.
------------------
PV Name:                Type:           Description:
di0:port0:in0   bi                      Digital Input 1
di0:port0:in1   bi                      Digital Input 2
di0:port0:in2   bi                      Digital Input 3
di0:port0:in3   bi                      Digital Input 4
di0:port0:in4   bi                      Digital Input 5
di0:port0:in5   bi                      Digital Input 6
di0:port0:in6   bi                      Digital Input 7
di0:port0:in7   bi                      Digital Input 8
di0:port1:in0   bi                      Digital Input 9
di0:port1:in1   bi                      Digital Input 10
di0:port1:in2   bi                      Digital Input 11
di0:port1:in3   bi                      Digital Input 12
di0:port1:in4   bi                      Digital Input 13
di0:port1:in5   bi                      Digital Input 14
di0:port1:in6   bi                      Digital Input 15
di0:port1:in7   bi                      Digital Input 16
di0:port2:in0   bi                      Digital Input 17
di0:port2:in1   bi                      Digital Input 18
di0:port2:in2   bi                      Digital Input 19
di0:port2:in3   bi                      Digital Input 20
di0:port2:in4   bi                      Digital Input 21
di0:port2:in5   bi                      Digital Input 22
di0:port2:in6   bi                      Digital Input 23
di0:port2:in7   bi                      Digital Input 24
di0:port3:in0   bi                      Digital Input 25
di0:port3:in1   bi                      Digital Input 26
di0:port3:in2   bi                      Digital Input 27
di0:port3:in3   bi                      Digital Input 28
di0:port3:in4   bi                      Digital Input 29
di0:port3:in5   bi                      Digital Input 30
di0:port3:in6   bi                      Digital Input 31
di0:port3:in7   bi                      Digital Input 32



RELEASE_NOTES
#==============================================================
#
#  Name: RELEASE_NOTES
#
#  Rem: This files describes release notes for
#       the Xycom 2445 driver and device support.
#
#  Auth: dd-mmm-yyyy, Andy Foster        (USERNAME)
#  Rev:  dd-mmm-yyyy, Reviewer's Name    (USERNAME)
#--------------------------------------------------------------
#  Mod:
#        20-Mar-2010, S. Hoobler         (SONYA):
#          release xy2445-R1-0-lcls3
#        23-Feb-2010, K. Luchini         (LUCHINI):
#          release xy2445-R1-0-lcls2
#        29-Apr-2009, K. Luchini         (LUCHINI):
#          release xy2445-R1-0-lcls1 
#        27-Apr-2009, K  Luchini         (LUCHINI):
#          release xy2445-R2-0-0
#        06-Nov-2008, K. Luchini         (LUCHINI):
#          release xy2445-R1-0-6
#        21-Feb-2008, K. Luchini         (LUCHINI):
#          release xy2445-R1-0-5
#        10-Dec-2007, K. Luchini         (LUCHINI):
#          release xy2445-R1-0-4
#        08-Oct-2007, K. Luchini         (LUCHINI):
#          release xy2445-R1-0-3
#          release xy2445-R1-0-2
#        30-Aug-2007, K. Luchini         (LUCHINI):
#          release xy2445-R1-0-1
#
#==============================================================
#
R2.8.5           Vincent Esposito 10/17/2022
                 - Using IPAC R2.15-1.0.2
                 - Compile for rhel7

R2.8.2           Kristi Luchini  10/18/2019
                 - Using IPAC R2.14-0.2.0

R2.6.0-1.1.0     Marcio Donadio, 08/01/2018
                 - Improved .gitignore
		 - ipac/R2.8-1.2.1 was only released to RTEMS. For LinuxRT and rhe6, we need
		   to use ipac/R2.8-1.2.2
		 - ip445Report() function was missing

R2.6.0-1.0.1     Bruce Hill, 02/13/2017
				 Changed to ipac/R2.8-1.2.1 as that's the ipac release that
				 supports the linux pciE cards

R2.6.0-1.0.0     Marcio Donadio, 02/13/2017
                 Applied patches direct-16bits-read-write.patch, ip445-signal-template-arch-req.patch, add_LIBRARY_IOC_Linux_to_Makefile.patch to release R2.6.0. These patches does the following:
                 - PCDS way of reading and writing 16 bits directly.
                 - add archive and autosave template request files for ip445 signals.
                 - add LIBRARY_IOC_Linux   = Xy2445 ip445 in src Makefile.

R2.6.0:          Marcio Donadio, 02/13/2017
                 - Included ip445_552c.pdf manual.
                 - Changed DTYP from ACROMAG-IP445 to IP445 in order to keep the same standard used in IP330, IP340, IP440. 

R2.5.0:          Marcio Donadio, 02/10/2017
                 Functions xy2445Create and ip2445Create were not included when
                 changing code to support both names: xy2445 and ip445.

R2.4.0:          Marcio Donadio, 02/09/2017
                 - Some files modified to build for git workflow.
                 - Updated lib, dbd, and other files to support ip445 vs xy2445.
                   The old xy2445 variants are still supported for backward compatibility.

R2.3.0:          Marcio Donadio, 02/07/2017
                 Creating new release R2.3.0 to avoid confusion w/ prior pcds R2.2.* releases
                 Also updated to new git workflow base build support

xy2445-R1-0_1-1: Build against ipac-R2-11_1-2 which builds for all targets
                 Build for target linuxRT-x86_64

xy2445-R1-0_1-0: Build against ipac-R2-11_1-0 which builds for all targets

xy2445-R1-0-lcls6: Build with ipac-R2-11-lcls1
01/19/2012         Add xipIo.h to compile with ipac-R2-11-lcls1
                   Add xioIo.c file and copy xipIoParse() and strdupn() from drvXy9660.c (ipac-R2-8-lcls8).

xy2445-R1-0-lcls5: Build against EPICS R3-14-12 and use ipac-R2-8-lcls8

xy2445-R1-0-lcls4: Changes to decouple the version of EPICS BASE from the configure/RELEASE
                   Everything is essentially done in the RELEASE_SITE file now.

xy2445-R1-0-lcls3: Build with base-R3-14-8-2-lcls-6
                   Build with rtems-4.9.4

xy2445-R1-0-lcls2: Build with base-R3-14-8-2-lcls-5
                   Build with rtems-4.9.3

xy2445-R1-0-lcls1: Build with base-R3-14-8-2-lcls-3
                   Build with rtems-4.9.1
                   Comments out readback during init to help restore
                   same as xy2445-R2-0-0, but use new naming convention
                   for drivers maintained outside of SLAC.

xy2445-R1-0-6:   Build with base-R3-14-8-2
                 Build with ipac-R2-8-lcls3
                 Change drvier to use volatile for ptrs to
                 onboard registers
                 
xy2445-R1-0-5:   Build with base-R3-14-9-lcls1
                 Build with ipac-R2-8-lcls4
                 Replace CONFIG_APP with CONFIG_SITE
                 Add RELEASE_SITE            

xy2445-R1-0-4:   Backed out epics base-R3-14-8-2-lcls2
                 Build with ipac-R2-8-lcls3

xy2445-R1-0-3:   Build with base-R3-14-9-lcls1
                 Build with extensions-R3-14-9
                 Build with ipac-R2-8-lcls2

xy2445-R1-0-2:   Build with base-R3-14-8-2-lcls2
                 Build with ipac-R2-8-lcls1

xy2444-R1-0-lcls1: Use naming convention for tag

xy2445-R1-0-1:   Build with ipac-R1-0-2 instead of ipac-R1-0-1

xy2445-R1-0-0:   Initial release


xy2445App/src/xipIo.c
/*******************************************************************************

Project:
    IndustryPack Driver Interface for EPICS

File:
    xipIo.c

Copied xipIoParse() from original drvXy9660.c file
*******************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#include "xipIo.h"


char *strdupn( const char *ct, size_t n )
{
  char *duplicate;

  duplicate = (char *)malloc(n+1);
  if( !duplicate )
    return NULL;

  memcpy(duplicate, ct, n);
  duplicate[n] = '\0';

  return duplicate;
}

int xipIoParse( char *str, xipIo_t *ptr, char flag )
{
  char *name;
  char *end;

  if( str == NULL || ptr == NULL )
    return 1;

  while( !isalnum(*str) )
  {
    if( *str++ == '\0')
      return 1;
  }

  name = str;

  str = strpbrk(str, " ");
  if( str == NULL || *str == '\0' )
    return 1;

  ptr->name = strdupn(name, str - name);
  if( ptr->name == NULL )
    return 1;

  if( flag == 'A' )
  {
    end = strchr(str,'C');
    if( end )
    {
      str = end + 1;
      sscanf(str,"%d", &ptr->channel);
    }
    else
      return 1;
  }
  else if( flag == 'B' )
  {
    end = strchr(str,'P');
    if( end )
    {
      str = end + 1;
      sscanf(str,"%d", &ptr->port);
      end = strchr(str,'B');
      if( end )
      {
        str = end + 1;
        sscanf(str,"%d", &ptr->bit);
      }
      else
        return 1;
    }
    else
      return 1;
  }
  return(0);
}


xy2445App/src/xipIo.h
/*******************************************************************************

Project:
    Gemini Multi-Conjugate Adaptive Optics Project

File:
    xipIo.h

Description:
    Header file defining the parsing of addresses for
    XIP modules.

Author:
    Andy Foster <ajf@observatorysciences.co.uk>

Created:
      12th November 2002

Copyright (c) 2002 Andy Foster

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*******************************************************************************/

#ifndef INCxipIoH
#define INCxipIoH

/* Error Numbers */

#ifndef M_xip
#define M_xip  (603 <<16)
#endif

#define S_xip_badAddress  (M_xip| 1) /*XIP address syntax error*/


typedef struct
{
  char          *name;
  int           port;
  int           bit;
  int           channel;
  unsigned char intHandler;
} xipIo_t;

/* Function Prototypes */

int   xipIoParse( char *str, xipIo_t *ptr, char flag );
char *strdupn( const char *ct, size_t n ); 

#endif  /* INCxipIoH */


xy2445App/src/drvXy2445.h
/*******************************************************************************

Project:
    Gemini Multi-Conjugate Adaptive Optics Project

File:
    drvXy2445.h

Description:
    Header file for the XIP-2445-000 Industrial I/O Pack
    32-Channel Isolated SSR Output Module

Author:
    Andy Foster <ajf@observatorysciences.co.uk>

Created:
    12th November 2002

Copyright (c) 2002 Andy Foster

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*******************************************************************************/

#ifndef INCdrvXy2445H
#define INCdrvXy2445H

#define MAXPORTS 4
#define MAXBITS  8

/* Data sizes that can be read/written */
#define BIT      0
#define NIBBLE   1
#define PORT     2
#define WORD     3

/* Error numbers */

#ifndef M_xy2445
#define M_xy2445  (601 <<16)
#endif

#define S_xy2445_duplicateDevice (M_xy2445| 1) /*Duplicate xy2445 device definition*/
#define S_xy2445_validateFailed  (M_xy2445| 2) /*Card not validated*/
#define S_xy2445_mallocFailed    (M_xy2445| 3) /*Malloc Failure*/
#define S_xy2445_portError       (M_xy2445| 4) /*Error in port number*/
#define S_xy2445_bitError        (M_xy2445| 5) /*Error in bit number*/
#define S_xy2445_readError       (M_xy2445| 6) /*Read error*/
#define S_xy2445_writeError      (M_xy2445| 7) /*Write error*/
#define S_xy2445_cardNotFound    (M_xy2445| 8) /*Card not found*/
#define S_xy2445_dataFlagError   (M_xy2445| 9) /*Error in Data Flag*/


/* Memory Map for the Xy2445 Binary Output Module */

struct map2445
{
  unsigned short cntl_reg;   /* control register */
  struct io
  {
    unsigned short io_port;  /* IO registers */
  } io_map[4];
};


/* Structure to hold the board's configuration information */

struct config2445
{
  struct config2445 *pnext;       /* to next device. Must be first member */
  char              *pName;       /* Name to identify this card           */
  unsigned short    card;         /* Number of IP carrier board           */
  unsigned short    slot;         /* Slot number in carrier board         */
  volatile struct map2445    *brd_ptr;     /* pointer to base address of board     */
  unsigned short    *id_ptr;      /* board ID Prom                        */
};

long          xy2445Report( int interest );
long          xy2445Initialise( void );
long          ip445Report( int interest );
long          ip445Initialise( void );
int           xy2445Create( char *pName, unsigned short card,
                            unsigned short slot );
int           ip445Create( char *pName, unsigned short card,
                            unsigned short slot );
void          xy2445SetConfig( char *pName, unsigned short card,
                               unsigned short slot,
                               struct config2445 *pconfig );
long          xy2445Read( char *name, short port, short bit, int readFlag,
                          unsigned short *pval, int debug );
long          xy2445Write( char *name, short port, short bit, int writeFlag,
                           long value, int debug );
void          *xy2445FindCard( char *name );

#endif  /* INCdrvXy2445H */


xy2445App/src/drvXy2445.c
/*******************************************************************************

Project:
    Gemini Multi-Conjugate Adaptive Optics Project

File:
    drvXy2445.c

Description:
    EPICS Driver for the XIP-2445-000 Industrial I/O Pack
    32-Channel Isolated SSR Output Module

Author:
    Andy Foster <ajf@observatorysciences.co.uk>

Created:
    12th November 2002

Copyright (c) 2002 Andy Foster

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*******************************************************************************/

#ifdef NO_EPICS
#include <vxWorks.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* These are the IPAC IDs for this module */
#define IP_MANUFACTURER_XYCOM 0xa3
#define IP_MODEL_XYCOM_2445   0x09

#include "drvIpac.h"
#include "drvXy2445.h"

#ifndef NO_EPICS
#include "devLib.h"
#include "drvSup.h"
#include "epicsExport.h"
#include "iocsh.h"

/* EPICS Driver Support Entry Table */

#ifndef USE_TYPED_DRVET
struct drvet drvXy2445 = {
  2,
  (DRVSUPFUN) xy2445Report,
  (DRVSUPFUN) xy2445Initialise
};
#else
const drvet drvXy2445 = {2, xy2445Report, xy2445Initialise};
#endif

epicsExportAddress(drvet, drvXy2445);

#ifndef USE_TYPED_DRVET
struct drvet drvip445 = {
  2,
  (DRVSUPFUN) ip445Report,
  (DRVSUPFUN) ip445Initialise
};
#else
const drvet drvip445 = {2, ip445Report, ip445Initialise};
#endif

epicsExportAddress(drvet, drvip445);
#endif

LOCAL struct config2445 *ptrXy2445First = NULL;

long ip445Report( int interest ) 
{ 
  return xy2445Report(interest);
}

long xy2445Report( int interest )
{
  int               i;
  int               j;
  unsigned short    val;
  struct config2445 *plist;

  plist = ptrXy2445First;
  while( plist )
  {
    if( interest == 0 || interest == 2 )
    {
      printf("\nBoard Status Information: %s\n", plist->pName);
      printf("\nBase IO Address:          0x%lx\n", (unsigned long)plist->brd_ptr);
      printf("\nIdentification:              ");
      for(i = 0; i < 4; i++)                 /* identification */
        printf("%c",plist->id_ptr[i]);
      printf("\nManufacturer's ID:           %x",(unsigned char)plist->id_ptr[4]);
      printf("\nIP Model Number:             %x",(unsigned char)plist->id_ptr[5]);
      printf("\nRevision:                    %x",(unsigned char)plist->id_ptr[6]);
      printf("\nReserved:                    %x",(unsigned char)plist->id_ptr[7]);
      printf("\nDriver I.D. (low):           %x",(unsigned char)plist->id_ptr[8]);
      printf("\nDriver I.D. (high):          %x",(unsigned char)plist->id_ptr[9]);
      printf("\nTotal I.D. Bytes:            %x",(unsigned char)plist->id_ptr[10]);
      printf("\nCRC:                         %x",(unsigned char)plist->id_ptr[11]);
      printf("\n\n");
    }

    if( interest == 1 || interest == 2 )
    {
      printf("\nBoard Pattern: %s\n", plist->pName);
      printf("--------------\n");
      printf("              Bits\n");
      printf("         0 1 2 3 4 5 6 7\n");
      printf("         - - - - - - - -\n");
      for( i=0; i<MAXPORTS; i++ )
      {
        printf("Port %d:  ", i);
        for( j=0; j<MAXBITS; j++ )
        {
          xy2445Read( plist->pName, i, j, BIT, &val, 0 );
          printf("%d ", val);
        }
        printf("\n");
      }
      printf("\n");
    }
    plist = plist->pnext;
  }
  return(OK);
}

long _ip445Initialise( const char * functionName )
{
  return(OK);
}

long xy2445Initialise( void )
{
  return _ip445Initialise( "xy2445Initialise" );
}

long ip445Initialise( void )
{
  return _ip445Initialise( "ip445Initialise" );
}

int _ip445Create(
	const char	*	functionName, 
	char 		*	pName, 
	unsigned short card, 
	unsigned short slot )
{
  struct config2445 *plist;
  struct config2445 *pconfig;
  int               status;

  status = ipmValidate(card, slot, IP_MANUFACTURER_XYCOM, IP_MODEL_XYCOM_2445);
  if( status )
  {
    printf("%s: Error %d from ipmValidate\n", functionName, status);
    return S_xy2445_validateFailed;
  }

  if( !ptrXy2445First )
  {
    ptrXy2445First = malloc(sizeof(struct config2445));
    if( !ptrXy2445First )
    {
      printf("%s: First malloc failed\n", functionName);
      return S_xy2445_mallocFailed;
    }
    else
      xy2445SetConfig( pName, card, slot, ptrXy2445First );
  }
  else
  {
    /* Check for unique card */

    plist = ptrXy2445First;
    while( TRUE )
    {
      if( !strcmp(plist->pName, pName) || ((plist->card == card) && (plist->slot == slot)) )
      {
        printf("%s: Duplicate device (%s, %d, %d)\n", functionName, pName, card, slot);
        return S_xy2445_duplicateDevice;
      }
      if( plist->pnext == NULL )
        break;
      else
        plist = plist->pnext;
    }

    /* plist now points to the last item in the list */

    pconfig = malloc(sizeof(struct config2445));
    if( pconfig == NULL )
    {
      printf("%s: malloc failed\n", functionName);
      return S_xy2445_mallocFailed;
    }

    /* pconfig is the configuration block for our new card */
    xy2445SetConfig( pName, card, slot, pconfig );

    /* Update linked-list */
    plist->pnext = pconfig;
  }
  return(OK);
}

int xy2445Create(
        char            *       pName,
        unsigned short card,
        unsigned short slot )

{
	return _ip445Create("xy2445Create", pName, card, slot);
}

int ip445Create(
        char            *       pName,
        unsigned short card,
        unsigned short slot )
{
	return _ip445Create("ip445Create", pName, card, slot);
}

void xy2445SetConfig( char *pName, unsigned short card, unsigned short slot,
                      struct config2445 *pconfig )
{
  pconfig->pnext   = NULL;
  pconfig->pName   = pName;
  pconfig->card    = card;
  pconfig->slot    = slot;
  pconfig->brd_ptr = (volatile struct map2445 *)ipmBaseAddr(card, slot, ipac_addrIO);
  pconfig->id_ptr = (unsigned short *) ipmBaseAddr(card, slot, ipac_addrID);
  
  /* Perform a software reset */
  pconfig->brd_ptr->cntl_reg = 0x01;
}


long xy2445Read( char *name, short port, short bit, int readFlag, 
                 unsigned short *pval, int debug )
{
  struct config2445 *plist;
  volatile struct map2445    *map_ptr;
  unsigned char     port0;
  unsigned char     port1;
  unsigned char     port2;
  unsigned char     port3;
  unsigned int      res;
  int               shift;

  if( (port < 0) || (port >= MAXPORTS) )
  {
    printf("xy2445Read: %s: port number out of range %d\n", name, port );
    return S_xy2445_portError;
  }
  else if( (bit < 0) || (bit >= MAXBITS) )
  {
    printf("xy2445Read: %s: bit number out of range %d\n", name, bit );
    return S_xy2445_bitError;
  }
  else
  {
    plist = xy2445FindCard( name );
    if( plist )
    {
      map_ptr = plist->brd_ptr;
      if( readFlag == BIT || readFlag == PORT )
      {
        *pval = (unsigned short) map_ptr->io_map[port].io_port;
        if( readFlag == BIT )
        {
          if( *pval & (1 << bit) )
            *pval = 1;
          else
            *pval = 0;
        }
      }
      else
      {
        port0 = map_ptr->io_map[0].io_port;
        port1 = map_ptr->io_map[1].io_port;
        port2 = map_ptr->io_map[2].io_port;
        port3 = map_ptr->io_map[3].io_port;

        /* Combine into a 32-bit integer */
        res   = (port3<<24) + (port2<<16) + (port1<<8) + port0;

        /* Calculate position in integer where we want to be */
        shift = port*MAXBITS + bit;

        if( readFlag == NIBBLE )
          *pval = (res >> shift) & 0xF;
        else if( readFlag == WORD )
          *pval = (res >> shift) & 0xFFFF;
        else
        {
          printf("xy2445Read: %s: Data flag error (%d)\n", name, readFlag );
          return S_xy2445_readError;
        }
      }
      if( debug )
        printf("xy2445Read: name = %s, port = %d, bit = %d, readFlag = %d, value = 0x%x\n", 
                name, port, bit, readFlag, *pval);
    }
    else
    {
      printf("xy2445Read: Card %s not found\n", name);
      return S_xy2445_cardNotFound;
    }
  }
  return(OK);
}


long xy2445Write( char *name, short port, short bit, int writeFlag,
                  long value, int debug )
{
  struct config2445 *plist;
  volatile struct map2445    *map_ptr;
  unsigned char     port0;
  unsigned char     port1;
  unsigned char     port2;
  unsigned char     port3;
  unsigned char     newport;
  unsigned char     bpos;
  unsigned int      res;
  unsigned int      zeroOut;
  unsigned short    zeroMask=0;
  int               shift;

  if( (port < 0) || (port >= MAXPORTS) )
  {
    printf("xy2445Write: %s: port number out of range %d\n", name, port );
    return S_xy2445_portError;
  }
  else if( (bit < 0) || (bit >= MAXBITS) )
  {
    printf("xy2445Write: %s: bit number out of range %d\n", name, bit );
    return S_xy2445_bitError;
  }
  else
  {
    plist = xy2445FindCard( name );
    if( plist )
    {
      if( debug )
        printf("xy2445Write: name = %s, port = %d, bit = %d, writeFlag = %d, value = 0x%lx\n", 
                             name, port, bit, writeFlag, value);
      map_ptr = plist->brd_ptr;
      if( writeFlag == BIT )
      {
        if( value < 0x0 || value > 0x1 )
        {
          printf("xy2445Write: %s: BIT value out of range = %ld\n", name, value);
          return S_xy2445_writeError;
        }
        else
        {
          bpos  = 1 << bit;
          value = value << bit;
          map_ptr->io_map[port].io_port = 
                        (int)((map_ptr->io_map[port].io_port & ~bpos) | value);
        }
      }
      else if( writeFlag == PORT )
      {
        if( value < 0x0 || value > 0xFF )
        {
          printf("xy2445Write: %s: PORT value out of range = %ld\n", name, value);
          return S_xy2445_writeError;
        }
        else
          map_ptr->io_map[port].io_port = value;
      }
      else if( writeFlag == NIBBLE || writeFlag == WORD )
      {
        if( (writeFlag == NIBBLE) && (value < 0x0 || value > 0xF) )
        {
          printf("xy2445Write: %s: NIBBLE value out of range = %ld\n", name, value);
          return S_xy2445_writeError;
        }
        else if( (writeFlag == WORD) && (value < 0x0 || value > 0xFFFF) )
        {
          printf("xy2445Write: %s: WORD value out of range = %ld\n", name, value);
          return S_xy2445_writeError;
        }
        else
        {
          /* Read all ports */
          port0 = map_ptr->io_map[0].io_port;
          port1 = map_ptr->io_map[1].io_port;
          port2 = map_ptr->io_map[2].io_port;
          port3 = map_ptr->io_map[3].io_port;
 
          /* Combine into a 32-bit unsigned integer */
          res = (port3<<24) + (port2<<16) + (port1<<8) + port0;

          shift  = port*MAXBITS + bit;
            
          /* We need to only change the Nibble/Word of the 32-bits */
          /* and leave the remaining bits unchanged                */
          if( writeFlag == NIBBLE )
            zeroMask = 0xF;
          else if(  writeFlag == WORD )
            zeroMask = 0xFFFF;

          /* Zero-out the bits we want to change */
          zeroOut = 0xFFFFFFFF & ~(zeroMask<<shift);

          /* Current value AND zeroOut will zero-out the bits.   */
          /* Now OR this with the new bit pattern shifted by the */
          /* appropriate amount                                  */

          if( debug )
            printf("xy2445Write:  res (old) = 0x%x, zeroOut = 0x%x, value = 0x%lx, shift = %d\n", 
                                  res, zeroOut, value, shift);

          res = (res & zeroOut) | (value<<shift);

          if( debug )
            printf("xy2445Write:  res (new) = 0x%x\n", res);

          /* Write new port values */
          newport = res & 0xFF;
          map_ptr->io_map[0].io_port = newport;
          newport = (res>>8) & 0xFF;
          map_ptr->io_map[1].io_port = newport;
          newport = (res>>16) & 0xFF;
          map_ptr->io_map[2].io_port = newport;
          newport = (res>>24) & 0xFF;
          map_ptr->io_map[3].io_port = newport;
        }
      }
      else
      {
        printf("xy2445Write: %s: Data flag error (%d)\n", name, writeFlag );
        return S_xy2445_dataFlagError;
      }
    }
    else
    {
      printf("xy2445Write: Card %s not found\n", name);
      return S_xy2445_cardNotFound;
    }
  }
  return(OK);
}


void *xy2445FindCard( char *name )
{
  struct config2445 *plist;
  int               foundCard = 0;

  plist = ptrXy2445First;
  while( plist )
  {
    if( !strcmp(plist->pName, name) )
    {
      foundCard = 1;
      break;
    }
    plist = plist->pnext;
  }

  if( !foundCard )
    return NULL;
  else
    return plist;
}

/*******************************************************************************
* EPICS iocsh Command registry
*/

#ifndef NO_EPICS

/* xy2445Report(int interest) */
static const iocshArg xy2445ReportArg0 = {"interest", iocshArgInt};
static const iocshArg * const xy2445ReportArgs[1] = {&xy2445ReportArg0};
static const iocshFuncDef xy2445ReportFuncDef =
    {"xy2445Report",1,xy2445ReportArgs};
static void xy2445ReportCallFunc(const iocshArgBuf *args)
{
    xy2445Report(args[0].ival);
}

/* xy2445Create( char *pName, unsigned short card, unsigned short slot ) */
static const iocshArg xy2445CreateArg0 = {"pName",iocshArgPersistentString};
static const iocshArg xy2445CreateArg1 = {"card", iocshArgInt};
static const iocshArg xy2445CreateArg2 = {"slot", iocshArgInt};
static const iocshArg * const xy2445CreateArgs[3] = {
    &xy2445CreateArg0, &xy2445CreateArg1, &xy2445CreateArg2 };
static const iocshFuncDef xy2445CreateFuncDef =
    {"xy2445Create",3,xy2445CreateArgs};
static void xy2445CreateCallFunc(const iocshArgBuf *arg)
{
    xy2445Create(arg[0].sval, arg[1].ival, arg[2].ival);
}

/* ip445Report(int interest) */
static const iocshArg ip445ReportArg0 = {"interest", iocshArgInt};
static const iocshArg * const ip445ReportArgs[1] = {&ip445ReportArg0};
static const iocshFuncDef ip445ReportFuncDef =
    {"ip445Report",1,ip445ReportArgs};
static void ip445ReportCallFunc(const iocshArgBuf *args)
{
    xy2445Report(args[0].ival);
}

/* ip445Create( char *pName, unsigned short card, unsigned short slot ) */
static const iocshArg ip445CreateArg0 = {"pName",iocshArgPersistentString};
static const iocshArg ip445CreateArg1 = {"card", iocshArgInt};
static const iocshArg ip445CreateArg2 = {"slot", iocshArgInt};
static const iocshArg * const ip445CreateArgs[3] = {
    &ip445CreateArg0, &ip445CreateArg1, &ip445CreateArg2 };
static const iocshFuncDef ip445CreateFuncDef =
    {"ip445Create",3,ip445CreateArgs};
static void ip445CreateCallFunc(const iocshArgBuf *arg)
{
    ip445Create(arg[0].sval, arg[1].ival, arg[2].ival);
}


LOCAL void drvXy2445Registrar(void) {
    iocshRegister(&xy2445ReportFuncDef,xy2445ReportCallFunc);
    iocshRegister(&xy2445CreateFuncDef,xy2445CreateCallFunc);
    iocshRegister(&ip445ReportFuncDef,ip445ReportCallFunc);
    iocshRegister(&ip445CreateFuncDef,ip445CreateCallFunc);
}
epicsExportRegistrar(drvXy2445Registrar);

LOCAL void drvip445Registrar(void) {
    iocshRegister(&xy2445ReportFuncDef,xy2445ReportCallFunc);
    iocshRegister(&xy2445CreateFuncDef,xy2445CreateCallFunc);
    iocshRegister(&ip445ReportFuncDef,ip445ReportCallFunc);
    iocshRegister(&ip445CreateFuncDef,ip445CreateCallFunc);
}
epicsExportRegistrar(drvip445Registrar);

#endif
